### ISSUE TITLE
Some comparative test reports

### ISSUE THREAD

**Issue by panxinmiao** [pygfx/pygfx#772]
### Background
We previously discussed potential performance improvements for pygfx. During this discussion, I made some assumptions that proved to be incorrect.

To verify some of these assumptions, I decided to conduct comparative tests, controlling variables to minimize interference.

Here are the results of my tests.

### Test Code

https://github.com/panxinmiao/wgpu-performance-test

The test scripts involves rendering a large number (15,000,000) of points (with sizes)  in various ways to measure FPS and compare results. The scenarios tested include:

- **Vertex Buffer** or **Storage Buffer**
- **interleaved buffer** or **multiple plain buffers**
- **Drawing shapes with math**, or **directly sampling shapes from textures** in the fragment shader.

The script names directly reflect the features used: 

- `points` indicates drawing standard dots, 
- `points_marker` indicates drawing "♣️" (logic from pygfx), 
- `points_marker_texture` indicates direct sampling from texture, 
- and `interleaved` indicates using an interleaved buffer for vertex or storage buffer.

#### Test Environment 1
- **OS:** Windows 11
- **Python:** 3.12
- **GPU:** NVIDIA GeForce RTX 3080 Ti
- **Driver:** 552.22
- **Backend:** Vulkan

| Script                                           | FPS   |
|--------------------------------------------------|-------|
| points_vertex_buffer                             | 64.7  |
| points_storage_buffer                            | 64.5  |
| points_interleaved_vertex_buffer                 | 19.9  |
| points_interleaved_storage_buffer                | 20.0  |
| points_marker_storage_buffer                     | 64.4  |
| points_marker_interleaved_storage_buffer         | 24.5  |
| points_marker_texture_interleaced_storage_buffer | 29.6  |

#### Test Environment 2
- **OS:** Windows 11
- **Python:** 3.11
- **GPU:** NVIDIA GeForce GTX 1080 Ti
- **Driver:** 531.68
- **Backend:** Vulkan

| Script                                           | FPS   |
|--------------------------------------------------|-------|
| points_vertex_buffer                             | 52.2  |
| points_storage_buffer                            | 63.7  |
| points_interleaved_vertex_buffer                 | 16.2  |
| points_interleaved_storage_buffer                | 16.1  |
| points_marker_storage_buffer                     | 60.8  |
| points_marker_interleaved_storage_buffer         | 18.1  |
| points_marker_texture_interleaced_storage_buffer | 22.4  |

### Conclusions

1. **Vertex Buffer vs. Storage Buffer**
   - On the RTX 3080 Ti, the FPS for `points_vertex_buffer` (64.7) and `points_storage_buffer` (64.5) are almost identical. On the RTX 1080 Ti, `storage_buffer` (63.7) is even outperforms `vertex_buffer` (52.2), 😅 
   - No significant difference was observed between `interleaved_vertex_buffer` and `interleaved_storage_buffer`.
   - Therefore, vertex_buffer and storage_buffer have almost no performance difference. While i heared that Storage Buffer might perform worse on mobile devices or older hardware, for pygfx, we don't need to consider these scenarios.

2. **Interleaved Buffer**
   - Contrary to my expectations, using interleaved_buffer resulted in significantly worse performance.
   - **RTX 3080 Ti**:
     - `points_vertex_buffer` (64.7) vs. `points_interleaved_vertex_buffer` (19.9)
     - `points_storage_buffer` (64.5) vs. `points_interleaved_storage_buffer` (20.0)
     - `points_marker_storage_buffer` (64.4) vs. `points_marker_interleaved_storage_buffer` (24.5)
   - **RTX 1080 Ti** shows consistent results.
   - This indicates that reading data from multiple buffers per vertex is much faster than reading from a single interleaved buffer. This is counterintuitive, as I thought the data layout of the interleaved buffer was specifically designed for vertex access, 😓 . 
   - However, in this use case, there is no data dependency when reading from two buffers, and it can be done in parallel.  I am not sure what the impact is when there is a data dependency (such as in pygfx, you need to first read the index buffer and then read other vertex attribute buffers after obtaining the results).
   - So what are the advantages of InterleavedBuffer? More compact data (due to byte alignment, Interleaved can more effectively utilize space) and reducing memory usage should be an advantage. Perhaps for dynamic data, it can also reduce the transmission consumption between CPU and GPU? 🤔 

3. **Drawing Shapes in Shader vs. Sampling from Texture**
   - **RTX 3080 Ti**:
     - `points_marker_texture_interleaced_storage_buffer` (29.6) vs. `points_marker_interleaved_storage_buffer` (24.5)
   - **RTX 1080 Ti**:
     - `points_marker_texture_interleaced_storage_buffer` (22.4) vs. `points_marker_interleaved_storage_buffer` (18.1)
   - Using texture sampling for drawing shapes is more performant than computing shapes in the shader. GPU texture sampling is highly optimized and fast, And this method is more flexible, we can draw points of any shape..

4. **Additional Observation**
   - Comparing `points_marker_interleaved_storage_buffer` (24.5) and `points_interleaved_storage_buffer` (20.0), the former surprisingly has higher FPS, even though it draws a more complex shape. This discrepancy persisted despite repeated tests.
   - After some consideration, I hypothesized that the difference is due to the varying area sizes of the shapes. With no depth (hence no Early-Z), both have significant overdraw. The difference in performance likely stems from the time taken to write results to the final texture.
   - Circular dots cover a larger area than "♣️" shapes. GPUs optimize alpha blending by discarding fragments with alpha 0 instead of writing them to the final texture. Since "♣️" shapes have a smaller area, they have more alpha 0 fragments.
   - To verify this, I disabled all blending (commented out the blend configuration in the pipeline) and re-ran the tests, In this case, no fragments should be discarded，and they will be written to the target texture, even if their alpha is 0.
      The result is in line with the prediction.

| Script                                           | RTX 3080 Ti | RTX 1080 Ti |
|--------------------------------------------------|-------------|-------------|
| points_vertex_buffer                             | 64.4        | 52.1        |
| points_storage_buffer                            | 64.5        | 55.7        |
| points_interleaved_vertex_buffer                 | 17.4        | 14.1        |
| points_interleaved_storage_buffer                | 17.3        | 14.0        |
| points_marker_storage_buffer                     | 64.2        | 52.9        |
| points_marker_interleaved_storage_buffer         | 17.2        | 13.5        |
| points_marker_texture_interleaced_storage_buffer | 21.8        | 17.2        |

- With blending disabled, `points_interleaved_storage_buffer` and `points_marker_interleaved_storage_buffer` show no significant difference. On the RTX 1080 Ti, circular dots are slightly faster (14.0 vs. 13.5).
- `points_marker_texture_interleaced_storage_buffer` remains the most performant.

---

### TL;DR:

- There is no significant performance difference between Vertex Buffer and Storage Buffer.
- Using InterleavedBuffer does not improve performance; instead, it significantly decreases access speed. (The possible benefit is saving video memory).
- In every aspect, using texture sampling to draw shapes is better than drawing shapes through mathematical calculations in the shader. (It's simpler, more efficient, and more flexible).
- Hardware-based alpha blending is extremely fast, with almost no additional overhead. In fact, enabling alpha blending allows the GPU to apply optimizations that reduce the number of fragments written to the target texture, thereby improving performance.

**Comment by Korijn** [comment]
Amazing work! Wow! Incredibly useful information. Well done!

**Comment by almarklein** [comment]
Wow, this is super helpful! Thanks!

> There is no significant performance difference between Vertex Buffer and Storage Buffer.

🎉 

> Using InterleavedBuffer does not improve performance; instead, it significantly decreases access speed. (The possible benefit is saving video memory).

This is surprising, especially the observation that it's actually slower 🤔 

> In every aspect, using texture sampling to draw shapes is better than drawing shapes through mathematical calculations in the shader. (It's simpler, more efficient, and more flexible).

I did not expect this! Though its good news, because we can use SDF (e.g. future `PointsSdfMaterial`) to render a wider variety of markers, without sacrificing performance. Or a `PointsCharMaterial` to render a Unicode symbol.

> Hardware-based alpha blending is extremely fast, with almost no additional overhead. In fact, enabling alpha blending allows the GPU to apply optimizations that reduce the number of fragments written to the target texture, thereby improving performance.

BTW the discarding of fragments with `alpha-0.0` happend because of this line in the shader:
https://github.com/pygfx/pygfx/blob/ef138442baf2b5f304a6980b1d530d85425ffbeb/pygfx/renderers/wgpu/wgsl/points.wgsl#L294-L296

Actually, the blending code does it too 😅 
https://github.com/pygfx/pygfx/blob/ef138442baf2b5f304a6980b1d530d85425ffbeb/pygfx/renderers/wgpu/engine/blender.py#L143-L150

**Comment by almarklein** [comment]
I ran the test scripts on Apple M1. I saw that the fps tends to fluctuate the first 10 secs or so. I took waited for it to stabilize and then averaged the last 10 reported fps. Seeing the same trends.

| Script                                           | FPS   |
|--------------------------------------------------|-------|
| points_vertex_buffer                             | 15.1  |
| points_storage_buffer                            | 17.9  |
| points_interleaved_vertex_buffer                 | 3.5  |
| points_interleaved_storage_buffer                | 3.3 |
| points_marker_storage_buffer                     | 17.1  |
| points_marker_interleaved_storage_buffer         | 2.9  |
| points_marker_texture_interleaced_storage_buffer | 3.5  |

**Comment by panxinmiao** [comment]
For the past couple of days, I've been searching for information on why the `InterleavedBuffer` might be slower, trying to understand the reason, but I couldn't find any relevant articles or discussions on the internet, which didn't seem reasonable. I suspected that I had made a mistake somewhere that led me to this conclusion. 

Just a moment ago, while discussing byte alignment issues in `wgpu-py`, I suddenly realized that the problem lay there – I had made an error. When the elements in the buffer are not Structures but basic elements, no additional `GPUCHUNCK_SIZE` alignment is needed. I modified the code and re-ran the tests, and now the results are consistent: the performance of `InterleavedBuffer` and plain buffers is almost indistinguishable (at least in these test examples). 

On RTX 3080 Ti:

| Script                                           | FPS   |
|--------------------------------------------------|-------|
| points_vertex_buffer                             | 19.7  |
| points_storage_buffer                            | 19.6  |
| points_interleaved_vertex_buffer                 | 19.8  |
| points_interleaved_storage_buffer                | 19.7  |
| points_marker_storage_buffer                     | 24.1  |
| points_marker_interleaved_storage_buffer         | 24.2  |
| points_marker_texture_interleaced_storage_buffer | 29.8  |

I'm sorry for providing some misleading information. 😅

**Comment by almarklein** [comment]
Ah, glad that we understand that result now too. Thanks for the update!
